### PR TITLE
HOCS-2301-"Unknown exception" exception only logs exception message

### PR DESCRIPTION
**Pull request by agraham19**
* add event and stacktrace to uncaught exceptions log output
* write tests for error handler

### CODE REVIEW

File: src/main/java/uk/gov/digital/ho/hocs/casework/domain/exception/RestResponseEntityExceptionHandler.java [hunk 1]
```diff
@@ -75,7 +79,10 @@ public ResponseEntity handle(HttpMessageNotReadableException e) {
 
     @ExceptionHandler(Exception.class)
     public ResponseEntity handle(Exception e) {
-        log.error("Exception: {}", e.getMessage(), value(EVENT, UNCAUGHT_EXCEPTION));
+        Writer stackTraceWriter = new StringWriter();
+        PrintWriter printWriter = new PrintWriter(stackTraceWriter);
+        e.printStackTrace(printWriter);
```
Comment (zuzak) [inline src/main/java/uk/gov/digital/ho/hocs/casework/domain/exception/RestResponseEntityExceptionHandler.java:84]: Does this get output as JSON? If not, it'll be a pain to dig out of Kibana

File: src/main/java/uk/gov/digital/ho/hocs/casework/domain/exception/RestResponseEntityExceptionHandler.java [hunk 2]
```diff
@@ -75,7 +79,10 @@ public ResponseEntity handle(HttpMessageNotReadableException e) {
 
     @ExceptionHandler(Exception.class)
     public ResponseEntity handle(Exception e) {
-        log.error("Exception: {}", e.getMessage(), value(EVENT, UNCAUGHT_EXCEPTION));
+        Writer stackTraceWriter = new StringWriter();
+        PrintWriter printWriter = new PrintWriter(stackTraceWriter);
```
Comment (parth-ho) [inline src/main/java/uk/gov/digital/ho/hocs/casework/domain/exception/RestResponseEntityExceptionHandler.java:83]: Could we not do something like `e.getStackTrace().toString()` instead? Not sure if we need to do this as I think the Exception object already has everything we need.

Comment (agraham19) [reply src/main/java/uk/gov/digital/ho/hocs/casework/domain/exception/RestResponseEntityExceptionHandler.java]: Unfortunately this just prints [Ljava.lang.StackTraceElement;@... I've done a quick Google, but there doesn't seem to be a simpler way to stringify the stacktrace.

File: src/main/java/uk/gov/digital/ho/hocs/casework/domain/exception/RestResponseEntityExceptionHandler.java (hunk 1 above)
Comment (agraham19) [reply src/main/java/uk/gov/digital/ho/hocs/casework/domain/exception/RestResponseEntityExceptionHandler.java]: The logback configuration should output the logs as json based on the profile used in dev.

File: src/main/java/uk/gov/digital/ho/hocs/casework/domain/exception/RestResponseEntityExceptionHandler.java [hunk 3]
```diff
@@ -75,7 +75,7 @@ public ResponseEntity handle(HttpMessageNotReadableException e) {
 
     @ExceptionHandler(Exception.class)
     public ResponseEntity handle(Exception e) {
-        log.error("Exception: {}", e.getMessage(), value(EVENT, UNCAUGHT_EXCEPTION));
+        log.error("An error occurred", e);
```
Comment (parth-ho) [inline src/main/java/uk/gov/digital/ho/hocs/casework/domain/exception/RestResponseEntityExceptionHandler.java:78]: We still need `value(EVENT, UNCAUGHT_EXCEPTION)` as that seems to be the way be define logs using this key value pair.

Comment (agraham19) [reply src/main/java/uk/gov/digital/ho/hocs/casework/domain/exception/RestResponseEntityExceptionHandler.java]: I have readded this.

**Review by parth-ho — APPROVED** [review]

**Pull request merged**